Translation and Validation for EBSCOhost

ThomasFleischmann opened the pull request:
This request entails:

- the search_query/parser_ebsco.py (fundamental functionalities for parsing)
- the search_query/parser_validation.py (fundamental functionalities for validating)
- the search_query/proximity_query_ebsco.py (A new operator like or_query.py or and_query.py
- the search_query/serializer_ebsco.py (translation from pre-notation to ebsco format)
- the search_query/save_file.py (written for the ebsco_example to safe files to ebsco format)
- the test/test_parser_ebsco.py (unit tests for tokenization, linter, and parsing)
- the test/ebsco_example.py (an example on how to load and translate queries, as well as how to build your own queries and save it to a json file)

geritwagner commented on search_query/parser_ebsco.py:
```diff
@@ -0,0 +1,474 @@
+        """Parse a query string."""
+
+        self.linter_messages.clear()
+        self.mode = "not-strict"
```
@ThomasFleischmann : should the self.mode always be set to "not-strict" here?

geritwagner commented:
Thank you! I have added unit tests and made a few revisions. If you agree with the current implementation of `add_artificial_parentheses_for_operator_precedence(...)`, we could share it with the others. We may even make it a method of `QueryStringParser` (which inheriting classes/parsers could use or override).

I also added a suggestion for handling linter messages. If you agree with the changes, we will need to add the validation messages to the constants (this will also make them available in the documentation).

geritwagner commented:
Note: We should also mention the `add_artificial_parentheses_for_operator_precedence(...)` in the developer documentation (as a method that can easily be reused).

geritwagner commented:
Thank you for having a look at the validation and for checking my suggestions. I would assume that the precedence could already be adapted by changing the `OPERATOR_PRECEDENCE` attribute (which is used in `get_precedence()`).

Adding the `add_artificial_parentheses()` to the skeleton in the docs would be great. If you like, you could also move the `add_artificial_parentheses()` to the `QueryStringParser` on the `main`branch (to make it available for the other parsers).

ThomasFleischmann commented:
These are the newest changes to the pull-request:

- changed linter handling as per suggestion
- removed unused linter list and messages in ebsco parser and linter
- added an ebsco-specific linter message (wrong token sequence)
- added some additional tests
- moved add_artificial_parentheses_for_precedence to parser base to be used by all parsers

Regarding the add_artificial_parentheses, i also made some small changes. Here i had some queries in the automated tests that came back erroneous. The reason was i believe this:   
natural* OR flood* OR earthquake* AND volcano* AND hurricane* NOT drought* OR tornado* OR climate* 
The current version added the correct amount of closed-parentheses, but not enough open ones before the NOT part of the query. 
I changed it so there are additional parentheses. 
This however means, that in some instances there are redundant parentheses, however they do not collide with its logic.

To be honest, i am unsure as to what the issue is.

geritwagner commented:
Thank you - I added the `flatten_redundant_artificial_nesting()`, which should resolve the issue. Considering that we have no "look-ahead" when adding artificial parentheses, adding them for each level of operator precedence (and removing redundant ones) seems to be the only viable option.

The last changes before we merge the pull request are to use the [Token enums](https://github.com/digital-work-lab/agenda-search-query-theses/issues/2). I assume the [test/ebsco_example.py](https://github.com/CoLRev-Environment/search-query/pull/19/files#diff-c5fda8856c589313c5465ab065e7012712ee3d17e80fb31f113628e56e61e722) could also be removed?

ThomasFleischmann commented:
Thank you for changing the enumeration. In the latest push i have only removed the test/ebsco_example.
